### PR TITLE
[https://nvbugs/6441022][fix] Enable CUDA graph for final context token computation

### PR DESCRIPTION
**Pull request by SimengLiu-nv**
## Summary by CodeRabbit

- **New Features**
  - Improved KV-cache reuse when only the final prompt token remains.
  - Added CUDA graph support for promoted final-context requests.
  - Preserved context logits, guided decoding, and multimodal position metadata during reuse.
  - Improved handling when reusable cache data is moved to secondary memory.

- **Bug Fixes**
  - Corrected CUDA graph selection and KV-cache restoration synchronization for reused context.

- **Tests**
  - Added unit and integration coverage for final-token reuse, guided decoding, multimodal prompts, and CUDA graph replay.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by nvpohanh** [comment]
[by Codex] @thorjohnsen Could you please review this PR? Thank you!

**Review by yufeiwu-nv — APPROVED** [review]

**Review by dpitman-nvda — APPROVED** [review]
Approval for infra

**Review by Shixiaowei02 — APPROVED** [review]

**Review by eopXD — APPROVED** [review]
Don't really see the purpose of adding test cases under `cpp/tests/unit_tests/batch_manager/kvCacheManagerTest.cpp` matters here. My opinion is we don't need them at all.

Promoting the context batch here to enable CUDA graph looks good to me.

It seems like the test cases won't get captured by how you specified them under the `.yml` files. I think you will need to update them.

Approving to unblock.

**Comment by SimengLiu-nv** [comment]
> Don't really see the purpose of adding test cases under `cpp/tests/unit_tests/batch_manager/kvCacheManagerTest.cpp` matters here. My opinion is we don't need them at all.
> 
> Promoting the context batch here to enable CUDA graph looks good to me.
> 
> It seems like the test cases won't get captured by how you specified them under the `.yml` files. I think you will need to update them.
> 
> Approving to unblock.

Hi @eopXD , thanks for the feedback. The purpose of the C++ tests is to pin the upstream KV-cache-manager contract that graph promotion depends on: final-token reuse must leave exactly one context token, preserve semantic request state/type and feature metadata, and correctly onboard shared offloaded blocks without breaking ownership or cancellation. Removing all of them would leave that cross-language boundary implicit.
I also rechecked the YAML selectors. The repository stage mapper recognizes all six entries: the five single-GPU tests map through l0_h100.yml to pre-merge PyTorch H100 stages, and the TP2 test maps through l0_dgx_h100.yml to DGX H100 PyTorch stages.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -3643,11 +3718,17 @@ def append_cross_attention_state(request: LlmRequest,
...
                 # the request has no previous tensor:
                 # (1) new_tokens_device is None, which means overlap scheduler is disabled; or
                 # (2) a dummy request; or
                 # (3) the first step in the generation server of disaggregated serving
-                if new_tokens_device is None or request.is_dummy or request.py_batch_idx is None:
+                if is_promoted_context:
+                    input_ids.append(
+                        request.get_tokens(0)[request.context_current_position])
+                    past_seen_token_num = request.context_current_position
+                elif new_tokens_device is None or request.is_dummy or request.py_batch_idx is None:
```
[the remark this answers is not in the document]
Comment (SimengLiu-nv) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:4483]: Fixed in new commit.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 2]
```diff
@@ -6106,14 +6212,36 @@ def forward(self,
...
+        promoted_context_request_ids: frozenset[int] = frozenset()
+        # TODO: Generalize these conservative gates as actual-draft, beam, and
+        # context-parallel providers for decoder-only LLMs gain support for
+        # promoted final-context rows. Each relaxation must preserve whole-batch
+        # fallback on graph miss and prove parity with the provider's native
+        # q_len=1 path. Encoder-decoder and non-LLM engines remain out of scope.
+        if (scheduled_requests.num_context_requests > 0
+                and self.cuda_graph_runner.enabled
+                and (not self.enable_spec_decode or
+                     (not self.is_draft_model and self.runtime_draft_len == 0))
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Seems there's a correctness issue for dynamic-tree speculative decoding: graph selection here use draft length 0, while input preparation later changes it to max_total_draft_tokens.  A mixed promoted-context/generation batch can then replay a zero-draft graph using nonzero-draft length inputs, causing a copy shape failure or incorrect metadata.

**Review by cascade812 — APPROVED** [review]
Overall, this looks good to me, and I’m approving to unblock.

One design consideration for the future: the PR places context requests in generation_requests and passes promoted_context_request_ids through several layers to preserve their original semantics. This works for the current change, though it may be somewhat fragile to maintain—future code may check only which list contains the request and accidentally apply generation behavior.

**Review by mikeiovine — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 2 above)
Comment (SimengLiu-nv) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Nice catch! Addressed in new commit.

**Review by StanleySun639 — APPROVED** [review]
